### ISSUE TITLE
Associate custom Python functions with a project instead of a user profile

### ISSUE THREAD

**Issue by captaincoordinates** [qgis/QGIS#39330]
**Feature description.**

Associate custom Python functions with a project instead of a user profile so that sharing a project automatically shares any custom functionality required for the proper functioning of that project.

**Additional context**

QGIS [documentation](https://docs.qgis.org/testing/en/docs/user_manual/working_with_vector/expression.html) states the following:
> Custom Python functions are stored under the user profile directory, meaning that at each QGIS startup, it will auto load all the functions defined with the current user profile. Be aware that new functions are only saved in the /python/expressions folder and not in the project file. If you share a project that uses one of your custom functions you will need to also share the .py file in the /python/expressions folder. 
This is problematic for projects that are authored by multiple users and/or managed by a version control system. Any remote updates to custom Python code must be explicitly copied to the user's expressions directory, and any local updates must be explicitly copied to the project's version-controlled directory. This issue will affect multiple authors working on the same map and a single author working across multiple machines.

QGIS should allow a project file to specify a relative path from which to load custom functions so that these can be version-controlled alongside the project file.

**Comment by andreasneumann** [comment]
I agree - that would be useful to have. Custom functions are often specific to a certain project (not always, but probably more often than not). Ideally one could choose whether to store custom functions in the profile or the project.

**Comment by pathmapper** [comment]
> Ideally one could choose whether to store custom functions in the profile or the project.

It would be useful to be able to specify a path other than or additional to the path of the user profile (e.g. a path to a shared network folder) where custom functions would be stored - like it is currently already possible for models and python scripts:

<img src="https://user-images.githubusercontent.com/20856381/95749592-16177680-0c9c-11eb-8d1e-9cef790d8657.png" width="200">

**Comment by nyalldawson** [comment]
It's not ideal but you CAN do this in current QGIS versions -- you just move the custom function code into the project's startup macro section

**Comment by zoner72** [comment]
> It's not ideal but you CAN do this in current QGIS versions -- you just move the custom function code into the project's startup macro section

This seems to work for me (using Pi), just had to enable macros on my colleagues computer but all showed up and worked. I agree its not ideal but at least we are keeping project specific code in the project it was written for.

thanks @nyalldawson

**Comment by denilsonjesus** [comment]
I tried but I couldn't Associate custom Python functions with a project instead of a user profile. Is there any other way to do what was suggested in the channel https://gis.stackexchange.com/questions/471136/filter-value-relation-dropdown-list-based-on-canvas-visibility-zoom-level-in- qgi/471141?noredirect=1#comment769973_471141 so that it works in any QGIS project?
​

**Comment by gkresic** [comment]
Isn't this solved by now? In QGIS 3.40.5 I can choose to add "Project Functions" by clicking a plus sign in the bottom left corner of the Function Editor and selecting "Project functions" in a drop-down list. I assume this also closes #20793, right?

**Comment by gacarrillor** [comment]
@gkresic, correct, and thanks for the heads up!
Closing as complete! Available since QGIS 3.40.

**Comment by mrpiggi** [comment]
Thanks for the update, I didn't notice it until now. Works like a charm, many thanks for the effort!
